src/libstd/panicking.rs

```diff
@@ -28,11 +30,131 @@ thread_local! {
     }
 }
 
-fn log_panic(obj: &(Any+Send), file: &'static str, line: u32,
-             log_backtrace: bool) {
-    let msg = match obj.downcast_ref::<&'static str>() {
+static HANDLER_LOCK: StaticMutex = MUTEX_INIT;
+static mut HANDLER: *const (Fn(&PanicInfo) + 'static + Sync + Send) = &default_handler;
+
+/// Unregisters the current panic handler, returning it.
+///
+/// If no custom handler is registered, the default handler will be returned.
+///
+/// # Panics
+///
+/// Panics if called from a panicking thread.
+#[unstable(feature = "panic_handler", reason = "awaiting feedback", issue = "30449")]
+#[cfg(not(stage0))] // Waiting on #28270 to hit snapshot
+pub fn take_handler() -> Box<Fn(&PanicInfo) + 'static + Sync + Send> {
+    if thread::panicking() {
+        panic!("cannot modify the panic handler from a panicking thread");
+    }
+
+    let _lock = HANDLER_LOCK.lock();
+    unsafe {
+        let handler = HANDLER;
+        HANDLER = &default_handler;
+        if handler == &default_handler as *const _ {
```

Comment:
In the interest of keeping the operations under a global lock as small as possible, could this conditional happen entirely outside of the unsafe block?